spec/requests.cddl

```diff
@@ -66,13 +66,21 @@ query-content = {
 query-response = tagged<{
   status: "replied"
   reply: call-reply
+  node_signatures: [* node-signature-of-query-response]
   //
   status: "rejected"
   reject_code: unsigned
   reject_message: text
   ? error_code: text
+  node_signatures: [* node-signature-of-query-response]
 }>
 
+node-signature-of-query-response = {
+  timestamp: timestamp
+  signature: bytes
+  identity: principal
+}
```

Comment:
This requires clients to separately fetch the public key from the IC, including figuring out the right subnet etc. This is different from update call signatures and from certified data, where the full certificate (including subnet delegation etc.) is included, and can be checked with no further data. Is there a reason to do that differently?

Reply:
I'd include a field `certificate` (like for certified variables) that reveals the corresponding public keys, and that it is responsible for the given canister.

Reply:
> This is different from update call signatures

It is still similar in the sense that in both cases you make a read_state request that gives you the delegation and the corresponding path in the subnet's certified state tree.

> Is there a reason to do that differently?

The main reason is to save network bandwidth by not including the certificate in every query response.

> I'd include a field certificate (like for certified variables) that reveals the corresponding public keys, and that it is responsible for the given canister.

This is indeed an alternative option to be discussed in the design review.

Reply:
Sounds like premature optimization at the expense of significant client complexity (caching, extra reques, staleness) to me.

My suggestion is:

* Allow clients to opt into signature in the request, to save work and bandwidth when the client doesn't care.
* Include the delegation and state tree in the response 
* include it in a way that with multiple replicas answering, the boundary node can merge the state trees (the data structure is set up for precisely such use cases!) to save bandwidth. I.e. a single value next to the list of signatures, not one in each.

Reply:
@nomeata in the design discussion few reasons were discussed in favour if not adding the certificate to the response.

1. It is not a premature optimization. [Node providers](https://wiki.internetcomputer.org/wiki/Node_Provider_Network_Setup_Guide) are required to have 300mbps  (37.5 MB/s) of bandwidth. All of this should account for all components that send and receive bytes over the wire - consensus, statesync, ingress, xnet, bitcoin, https outcalls, queries, read state, etc. When you start adding the required bandwidth for all those components once the IC scales just a bit more we start running into problems pretty soon.

In production queries right now are the biggest bandwidth consumer. Adding the certificate will more than double the query response. 

2. Adding the certificate in the response enables clients to do the same thing in two different ways which is not ideal for an API as well.

3. User complexity is pretty bounded given that only agents will have to implement this extra complexity, it won't impact the average user.

Reply:
Fair enough

Reply:
I got some stats of the certificate size for node keys.
In a network of 40 subnets and in the subnet with 40 nodes:
The certificate for 1 node key is around 1100 bytes. 
The certificate for 14 node keys varies from 1800 bytes to 2400 bytes. 
The certificate for 40 node keys is around 3300 bytes.
In comparison, a delegation is around 500 bytes.

@rumenov 
> In production queries right now are the biggest bandwidth consumer. 
I think there are two aspects regarding the bandwidth consume for certified queries.
1.  Will certified queries account for majority of queries in the future or used in cases where security is a big concern(e.g NNS dapp)?  
This is the same point I bring up regarding whether to make signatures optional.

2. Does fetching node keys through read_state requests really save the bandwidth?
Agents do not know which nodes will respond and thus need to fetch and cache all keys.
Agents do not know the subnet id when making query calls. It is not trivial to tell if agents already have node keys in cache and if a read_state call should be made.

Reply:
> 2. Does fetching node keys through read_state requests really save the bandwidth?
> [...]
> It is not trivial to tell if agents already have node keys in cache and if a read_state call should be made.

This is an excellent point. I guess it would be possible to simply invalidate the cache if the certificate is, e.g., older than a couple of minutes (it contains a timestamp). Not claiming this is trivial, but even if it would be, still
- many clients may not implement caching at all and so would do `read_state` requests on every query,
- many clients may only occasionally make queries meaning that that even if they implement caching they will have to re-fetch because the cache is already invalid at the point where they make the next query.

My takeaway from this is that there are use cases where the approach to fetch certificates separately via `read_state` is better and there are use cases where the approach to include certificates in the response is better. Or in other words, bandwidth consumption is probably not the deciding factor regarding whether or not to include the certificate in the reply. 

Taking this into account, I think there is no longer a good reason why we shouldn't allow the client to choose whether or not it wants to include the certificate in the response. This means going for option 2 from the [other thread](https://github.com/dfinity/interface-spec/pull/163#discussion_r1231842381) also including an option to request or not request a certificate seems the most reasonable to me.